File: todo-api/src/main/java/com/example/todoapi/service/TodoService.java
package com.example.todoapi.service;

import com.example.todoapi.model.SortBy;
import com.example.todoapi.model.SortDir;
import com.example.todoapi.model.Todo;
import com.example.todoapi.model.TodoStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.stereotype.Service;

import java.time.Instant;
import java.util.Comparator;
import java.util.List;
import java.util.NoSuchElementException;
import java.util.Optional;
import java.util.UUID;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Stream;

@Service
public class TodoService {

    private static final Logger log = LoggerFactory.getLogger(TodoService.class);

    private final ConcurrentHashMap<UUID, Todo> store = new ConcurrentHashMap<>();

    public List<Todo> findAll(TodoStatus status, String q, SortBy sortBy, SortDir sortDir) {
        Stream<Todo> stream = store.values().stream();

        stream = switch (status) {
            case ACTIVE -> stream.filter(t -> !t.isCompleted());
            case COMPLETED -> stream.filter(Todo::isCompleted);
            case ALL -> stream;
        };

        if (q != null && !q.isBlank()) {
            String lowerQ = q.trim().toLowerCase();
            stream = stream.filter(t ->
                    t.getTitle().toLowerCase().contains(lowerQ)
                    || (t.getDescription() != null && t.getDescription().toLowerCase().contains(lowerQ))
            );
        }

        Comparator<Todo> comparator = switch (sortBy) {
            case TITLE -> Comparator.comparing(t -> t.getTitle().toLowerCase());
            case CREATED_AT -> Comparator.comparing(Todo::getCreatedAt);
        };
        if (sortDir == SortDir.DESC) {
            comparator = comparator.reversed();
        }

        List<Todo> result = stream.sorted(comparator).toList();
        log.debug("findAll returned {} todos (status={}, q='{}', sortBy={}, sortDir={})",
                result.size(), status, q, sortBy, sortDir);
        return result;
    }

    public Optional<Todo> findById(UUID id) {
        log.debug("findById: {}", id);
        return Optional.ofNullable(store.get(id));
    }

    public Todo create(String title, String description) {
        Todo todo = new Todo();
        todo.setId(UUID.randomUUID());
        todo.setTitle(title);
        todo.setDescription(description);
        todo.setCompleted(false);
        Instant now = Instant.now();
        todo.setCreatedAt(now);
        todo.setUpdatedAt(now);
        store.put(todo.getId(), todo);
        log.info("Created todo: id={}, title='{}'", todo.getId(), todo.getTitle());
        return todo;
    }

    public Todo update(UUID id, String title, String description, boolean completed) {
        Todo todo = store.get(id);
        if (todo == null) throw new NoSuchElementException();
        todo.setTitle(title);
        todo.setDescription(description);
        todo.setCompleted(completed);
        todo.setUpdatedAt(Instant.now());
        log.info("Updated todo: id={}, completed={}", id, completed);
        return todo;
    }

    public Todo patch(UUID id, boolean completed) {
        Todo todo = store.get(id);
        if (todo == null) throw new NoSuchElementException();
        todo.setCompleted(completed);
        todo.setUpdatedAt(Instant.now());
        log.info("Patched todo: id={}, completed={}", id, completed);
        return todo;
    }

    public void delete(UUID id) {
        if (!store.containsKey(id)) throw new NoSuchElementException();
        store.remove(id);
        log.info("Deleted todo: id={}", id);
    }
}

File: todo-ui/src/app/core/services/todo.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Todo, CreateTodoDto, UpdateTodoDto, TodoFilter } from '../models/todo.model';

@Injectable({ providedIn: 'root' })
export class TodoService {
  private readonly http = inject(HttpClient);
  private readonly API = '/api/todos';

  findAll(filter?: TodoFilter): Observable<Todo[]> {
    let params = new HttpParams();
    if (filter?.status) params = params.set('status', filter.status);
    if (filter?.q?.trim()) params = params.set('q', filter.q);
    if (filter?.sortBy) params = params.set('sortBy', filter.sortBy);
    if (filter?.sortDir) params = params.set('sortDir', filter.sortDir);
    return this.http.get<Todo[]>(this.API, { params });
  }

  create(dto: CreateTodoDto): Observable<Todo> {
    return this.http.post<Todo>(this.API, dto);
  }

  update(id: string, dto: UpdateTodoDto): Observable<Todo> {
    return this.http.put<Todo>(`${this.API}/${id}`, dto);
  }

  patch(id: string, completed: boolean): Observable<Todo> {
    return this.http.patch<Todo>(`${this.API}/${id}`, { completed });
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.API}/${id}`);
  }
}
